recon/src/recon/mod.rs

```diff
@@ -79,46 +79,77 @@ impl Recon {
     }
 
     /// Generate a response message for a incoming message
-    pub fn process_message<H: Hash>(&mut self, received: &Message<H>) -> Message<H> {
+    pub fn process_message<H: Hash>(&mut self, received: &Message<H>) -> Response<H> {
+        let mut response = Response {
+            is_synchronized: true,
+            ..Default::default()
+        };
         // self.keys.extend(received.keys.clone()); // add received keys
         for key in &received.keys {
             let h = AHash::digest(key);
-            self.keys.insert(key.to_string(), h);
+            if self.keys.insert(key.to_string(), h).is_none() {
+                response.is_synchronized = false;
+            }
         }
         if self.keys.is_empty() {
-            return Message::<H>::default();
+            return response;
         }
         let mut received_keys = received.keys.iter();
         let mut received_hashs = received.ahashs.iter();
 
-        let mut response = Message::<H>::default(); // init the response empty
-
         let mut left_fencepost = self.keys.first_key_value().unwrap().0;
         let mut right_fencepost = received_keys
             .next()
             .unwrap_or(self.keys.last_key_value().unwrap().0);
         let mut received_hash = &H::identity();
         let zero = &H::identity();
 
-        response.keys.push(left_fencepost.to_string());
+        response.msg.keys.push(left_fencepost.to_string());
         while !received.keys.is_empty() && left_fencepost < received.keys.last().unwrap() {
-            response.process_range(left_fencepost, right_fencepost, received_hash, self);
+            response.is_synchronized &=
+                response
+                    .msg
+                    .process_range(left_fencepost, right_fencepost, received_hash, self);
             left_fencepost = right_fencepost;
             right_fencepost = received_keys
                 .next()
                 .unwrap_or(self.keys.last_key_value().unwrap().0);
             received_hash = received_hashs.next().unwrap_or(zero);
         }
         if !received.keys.is_empty() {
-            response.process_range(
+            response.is_synchronized &= response.msg.process_range(
                 received.keys.last().unwrap(),
```

Comment:
eventually we will need to plumb the interest range here rather than syncing the whole ring.